launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java

```diff
@@ -0,0 +1,335 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.apache.spark.launcher.CommandBuilderUtils.*;
+
+/**
+ * Special launcher for handling a CLI invocation of SparkSubmit.
+ * <p/>
+ * This launcher extends SparkLauncher to add command line parsing compatible with
+ * SparkSubmit. It handles setting driver-side options and special parsing needed
+ * for the different specialClasses.
+ * <p/>
+ * This class has also some special features to aid PySparkLauncher.
+ */
+class SparkSubmitCommandBuilder extends AbstractCommandBuilder {
+
+  /**
+   * Name of the app resource used to identify the PySpark shell. The command line parser expects
+   * the resource name to be the very first argument to spark-submit in this case.
+   *
+   * NOTE: this cannot be "pyspark-shell" since that identifies the PySpark shell to SparkSubmit
+   * (see java_gateway.py), and can cause this code to enter into an infinite loop.
+   */
+  static final String PYSPARK_SHELL = "pyspark-shell-main";
```

Comment:
Launching pythong directly would mean keeping shell code to separate pyspark argument from spark-submit arguments. That's needed to maintain compatibility because you can still launch python scripts through `bin/pyspark`.

If you want to break compatibility, then you don't need to go through Java to build the command, since all command line arguments would be spark-submit arguments, and handling arguments would be a lot easier.